test/Algebra/Graph/Test/RewriteRules.hs

```diff
@@ -13,111 +14,352 @@ module Algebra.Graph.Test.RewriteRules where
-overlays1, overlaysR :: [Graph a] -> Graph a
-overlays1 = transpose . overlays
-overlaysR = overlays . map transpose
+-- biclique
+bicliqueCP :: b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> Build a -> Build a -> b
+bicliqueCP e v o c xs ys = foldg e v o c (biclique (build xs) (build ys))
 
-inspect $ 'overlays1 === 'overlaysR
+inspect $ 'bicliqueCP `hasNoType` ''[]
+inspect $ 'bicliqueCP `hasNoType` ''Graph
 
-connects1, connectsR :: [Graph a] -> Graph a
-connects1 = transpose . connects
-connectsR = fromMaybe Empty . foldr (maybeF (flip Connect) . transpose) Nothing
+-- replaceVertex
+replaceVertexCP :: Eq a => a -> a ->
+  b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> Buildg a -> b
+replaceVertexCP u v e v' o c g =
+  foldg e v' o c (replaceVertex u v (buildg g))
 
-inspect $ 'connects1 === 'connectsR
+inspect $ 'replaceVertexCP `hasNoType` ''Graph
 
-vertices2 :: [a] -> Graph a
-vertices2 = transpose . overlays . map vertex
+-- mergeVertices
+mergeVerticesCP :: (a -> Bool) -> a ->
+  b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> Buildg a -> b
+mergeVerticesCP p v e v' o c g =
+  foldg e v' o c (mergeVertices p v (buildg g))
 
-inspect $ 'vertices2 === 'vertices1
+inspect $ 'mergeVerticesCP `hasNoType` ''Graph
 
-inspect $ 'vertices2 === 'verticesR
+-- splitVertex
+splitVertexCP :: Eq a => a -> Build a ->
+  b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> Buildg a -> b
+splitVertexCP x us e v o c g = foldg e v o c (splitVertex x (build us) (buildg g))
 
-cliqueT1, cliqueTR :: [a] -> Graph a
-cliqueT1 = transpose . connects . map vertex
-cliqueTR = fromMaybe Empty . foldr (maybeF (flip Connect) . Vertex) Nothing
+inspect $ 'splitVertexCP `hasNoType` ''[]
+inspect $ 'splitVertexCP `hasNoType` ''Graph
 
-inspect $ 'cliqueT1 === 'cliqueTR
+-- transpose
+transposeCP ::
+  b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> Buildg a -> b
+transposeCP e v o c g = foldg e v o c (transpose (buildg g))
 
-starT1, starTR :: a -> [a] -> Graph a
-starT1 x = transpose . star x
-starTR a [] = vertex a
-starTR a xs = connect (vertices xs) (vertex a)
+inspect $ 'transposeCP `hasNoType` ''Graph
 
-inspect $ 'starT1 === 'starTR
+-- simplify
+simple :: Eq g => (g -> g -> g) -> g -> g -> g
+simple op x y
+    | x == z    = x
+    | y == z    = y
+    | otherwise = z
+  where
+    z = op x y
 
-fmapFmap1, fmapFmapR :: Graph a -> (a -> b) -> (b -> c) -> Graph c
-fmapFmap1 g f h = fmap h (fmap f g)
-fmapFmapR g f h = fmap (h . f) g
+simplifyC, simplifyCR :: Ord a => Buildg a -> Graph a
+simplifyC  g = simplify (buildg g)
+simplifyCR g = g Empty Vertex (simple Overlay) (simple Connect)
 
-inspect $ 'fmapFmap1 === 'fmapFmapR
+inspect $ 'simplifyC === 'simplifyCR
 
-bind2, bind2R :: (a -> Graph b) -> (b -> Graph c) -> Graph a -> Graph c
-bind2 f g x = x >>= f >>= g
-bind2R f g x = x >>= (\x -> f x >>= g)
+-- compose
+composeC, composeCR :: Ord a => Buildg a -> Buildg a -> Graph a
+composeC  x y = compose (buildg x) (buildg y)
+composeCR x y = overlays
```

Comment:
Ouch. It's a bit sad we need to copy this.

Reply:
I agree but I don't see a better option here...